File: Bootloader/Common/src/image_util/image_download.c
/*
 *    Copyright 2021 CSEM SA
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
*/

#include "image_util/image_util.h"
#include <stdlib.h>
#include "hal/hal.h"

extern uint32_t _stImgHeader;
extern uint32_t _stImg2Header;
extern uint32_t _lnImg;

/**
 * @brief Downloads an image at the given address using UART
 * Intended to be used with the image_upload.py script
 *
 * @param img_header address of the recipient flash (i.e., should be the start of an image header)
 *
 * @return RET_OK on success, RET_ERROR on error
 */
static int download_image(uint32_t* img_header);

#ifdef ENABLE_TEST
// Only allow download to primary slot if tests are enabled
int download_image_primary(){
	return download_image(&_stImgHeader);
}
#endif /* ENABLE_TEST */


int download_image_secondary(){
	return download_image(&_stImg2Header);
}

/**
 * @brief Calls a function and return an error if its return value is != RET_OK
 *
 * @param func The function call, usable as in "func;"
 */
#define RETURN_IF_ERROR(func)			do { \
											int error = func; \
											if (error != RET_OK) { return error; } \
										}while(0)

static int download_image(uint32_t* img_header){

	uint32_t buf[HAL_FLASH_NB_32BITWORD_IN_FLASHWORD];
	char ack;

	// Wait for a ":" character before interpreting data
	do {
		RETURN_IF_ERROR(HAL_UART_RECEIVE((uint8_t*)&ack, 1));
	} while (ack != ':');
	
	uint32_t size = 0;

	// Ask the size
	RETURN_IF_ERROR(HAL_UART_RECEIVE((uint8_t*)&size, 4));

	if (size > (uint32_t)&_lnImg){
		return RET_ERROR;
	}

	// Erase flash that will hold the data
	RETURN_IF_ERROR(minimal_flash_erase_range((uint32_t)img_header, size));

	uint32_t processed=0;

	// Proceed flash word by flash word
	while(processed < size && processed < (uint32_t)&_lnImg){
		// Signal that we are ready to receive a flash word
		HAL_UART_PRINTF_NOLN(":");

		// Receive an entire flash word
		RETURN_IF_ERROR(HAL_UART_RECEIVE((uint8_t*)&buf, HAL_FLASH_WORD_SIZE));

		// Write it to flash
		RETURN_IF_ERROR(minimal_flash_write((uint32_t)img_header + processed, &buf, HAL_FLASH_WORD_SIZE));

		processed += HAL_FLASH_WORD_SIZE;
	}

	return RET_OK;
}

File: Bootloader/scripts/image_download.py
# 
#     Copyright 2021 CSEM SA
# 
#     Licensed under the Apache License, Version 2.0 (the "License");
#     you may not use this file except in compliance with the License.
#     You may obtain a copy of the License at
# 
#         http://www.apache.org/licenses/LICENSE-2.0
# 
#     Unless required by applicable law or agreed to in writing, software
#     distributed under the License is distributed on an "AS IS" BASIS,
#     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#     See the License for the specific language governing permissions and
#     limitations under the License.
# 

import time
import os
import serial
import sys
import serial.tools.list_ports
from typing import NamedTuple
from config.config import *

# Receive an image from UART and store it in a file
# @param image_name the name of the file in which the image will be stored
# @param ser the serial handler
# @param verbose whether prints are enabled
def receive_image(image_name, ser, verbose=False):

	f = open(image_name, 'wb')

	if verbose:
		print("contacting device...")

	ack = ':'
	ser.write(ack.encode('utf-8'))

	data = ser.read(8)
	size = int(data, 16)

	if verbose:
		print("received metadata, size=%d" % size)
		print("receiving image...")

	received = 0
	while(received < size):

		#signal we want a flash word
		ack = ':'
		ser.write(ack.encode('utf-8'))

		for i in range(8):
			data = ser.read(8)

			number = int(data, 16)
			#print (number)
			f.write(number.to_bytes(4, byteorder='big'))
			received += 4

			if verbose and received % 1024 == 0:
				print(str(((10000*received)//size)/100)+"%")

	if verbose:
		print("\ndone!")

# If the function is used as a stand-alone tool, define a main function
if __name__ == "__main__":

	ports = list(serial.tools.list_ports.comports())
	found = False

	assert len(sys.argv) >= 2, "an argument (output image file) is required!"

	for p in ports:
		if ("STLINK-V3" in p[1]):
			found = True
			ser = serial.Serial(p[0], baudrate = baudrate, parity=serial.PARITY_NONE, stopbits=serial.STOPBITS_ONE, bytesize=serial.EIGHTBITS)

			print(p[0]+" was detected as a valid port!")
			
			receive_image(sys.argv[1], ser, True)
			
			break

	if (not found):
		print("No valid port detected!")